### PR TITLE
*: Introduce data types

### PR DESCRIPTION
**Pull request by mxinden**
Modbus operates on 16 bit registers. These opaque bit streams can be
interpreted as different data types. This patch adds support for the
basic ones to be configured per metric.

In order to achieve the above, this patch also includes a couple
refactorings:

- Remove lexer and parser, instead configure the exporter with idiomatic
YAML, without a custom language on top. For example:

  ```
  targets:
    - name: simulator
      port: 127.0.0.1:1502
      id: 1
      analogIn:
        - name: register_twenty_two
          address: 22
          dataType: int16
  ```

- Rename `config.Slave` to `config.Target`.

### CODE REVIEW

**Comment by RichiH** [comment]
LGTM

File: modbus_exporter.go [hunk 1]
```diff
@@ -55,16 +54,19 @@ func main() {
...
 	}
 
-	parsedSlaves, err := parser.ParseSlaves(slavesFile)
-	if err != nil {
-		log.Fatalln(err)
-	}
-	err = modbus.RegisterData(parsedSlaves, slavesFile)
+	log.Info(config)
+
+	// TODO: Might be worth reusing some of the validation of ParseSlaves.
```
Comment (RichiH) [inline modbus_exporter.go:64]: Some more slave stuff to get rid of.

File: config/yaml.go [hunk 2]
```diff
@@ -26,32 +27,159 @@ import (
...
+		{
+			i := binary.BigEndian.Uint16(rawData[:])
+			return float64(int16(i)), nil
+		}
+	case ModbusUInt16:
+		{
+			i := binary.BigEndian.Uint16(rawData[:])
+			return float64(i), nil
+		}
+	case ModbusBool:
```
Comment (mxinden) [inline config/yaml.go:116]: @RichiH I have added a prove-of-concept implementation of the _boolean_ data type. Users specify a _register_ and a _bit position_ within the register.

Multiple questions arise on how to address bit positions within a register (16 bit):

-  Should the _bit position_ be zero indexed?

- Should the _bit position_ start with the least or most significant bit?

- What endianness do we use here?

Can you post a couple of examples with _16 bit input_, _bit position_ and _expected output_ here?

File: config/yaml_test.go [hunk 3]
```diff
@@ -0,0 +1,120 @@
...
+	two := 2
+
+	tests := []struct {
+		name          string
+		input         func() [2]byte
+		metricDef     func() *MetricDef
+		expectedValue float64
+	}{
+		{
+			name: "bool, no bit",
```
Comment (mxinden) [inline config/yaml_test.go:56]: @RichiH here you see couple of examples of the current _boolean data type_ implementation.

File: config/yaml.go (hunk 2 above)
Comment (RichiH) [reply config/yaml.go:116]: I think the only valid interpretation of these 16 bits is as a bit vector, which also sidesteps/answers these nicely. It's

* zero indexed (think of it as an offset to the base address)
* network order (first bit on the wire is 0)
* network order implies big endian, but this is not "valid", or meaningful, for bit vectors anyway.

This approach also matches most closely with what I saw in the field.

Given the above, do you still need samples?

File: config/yaml_test.go (hunk 3 above)
Comment (RichiH) [reply config/yaml_test.go:56]: Yes, this looks sane. `input[1] = uint8(1)` should then have `1` at `BitPosition: 9` aka zero-indexed 8.

**Pull request merged**